### PR TITLE
Update MultiplexedTensoredU2Box Circuit Construction

### PR DESCRIPTION
**Pull request by sjdilkes**
By interleaving the `{CX,U1}` and `{CX, Rz}` subcircuits comprising the `MultiplexedTensoredU2Box` Circuit construction, we can improve the Circuit depth without changing the number of gates.

This PR updates the `MultiplexedRotationBox::decompose` and `MultiplexedU2Box::decompose` methods to return, essentially, lists of relevant commands instead of modifying a `Circuit` object passed by reference.

For `MultiplexedRotationBox::generate_circuit` and `MultiplexedU2Box::generate_circuit` a Circuit object is constructed by iterating through these commands and adding to a `Circuit` object.

In `MultiplexedTensoredU2Box`, the construction is broken down into a series of `MultiplexedU2Box`, each synthesised separately. Now, when these new lists are returned from `decompose`, by synthesising each `MultiplexedU2Box` individually, the gates at the same index for each list are added at the same time. Additionally, the control conditions for each `MultiplexedU2Box` are permuted s.t. the order of CX gates for the `{CX,U1}` and `{CX, Rz}` subcircuits can differ between each construction. 

N.B. This is a draft PR as there are a couple of edge cases that generate the wrong unitary that I can't pinpoint the issue with. I thought some time and placating the CI might help me clock the issue.  *I wasn't initialising a float to 0 and an error threshold was too high.

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented hard-to-understand parts of my code.
- [x] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: tket/src/Circuit/Multiplexor.cpp [hunk 1]
```diff
@@ -860,59 +922,218 @@ Op_ptr MultiplexedTensoredU2Box::from_json(const nlohmann::json &j) {
...
+  for (unsigned i = 0; i < reference_size; i++) {
+    for (unsigned target = 0; target < all_decomps.size(); target++) {
+      GateSpec gate = all_decomps[target][i];
+      unsigned rotated_index;
+      switch (gate.type) {
+        case OpType::CX:
+          // we also need to map gate.qubit to the correct qubit
+          // we know that the bitstrings for the "target"th target have been
+          // left rotated by "target", so:
+          rotated_index = (*gate.qubit + (target % n_targets_)) % n_controls_;
```
Comment (yao-cqc) [inline tket/src/Circuit/Multiplexor.cpp:1028]: I'm a little confused by this...

File: tket/src/Circuit/Multiplexor.cpp [hunk 2]
```diff
@@ -860,59 +922,218 @@ Op_ptr MultiplexedTensoredU2Box::from_json(const nlohmann::json &j) {
...
-      // update the diagonal on the control qubits
-      diag_vec[j] *= p;
     }
-    if (!multip_rz.empty()) {
-      diag_circ.add_box(MultiplexedRotationBox(multip_rz), args);
+  }
+  return;
+}
+
+Eigen::VectorXcd combine_diagonals(
```
Comment (yao-cqc) [inline tket/src/Circuit/Multiplexor.cpp:1045]: Can you explain how this `rotate` works?

File: tket/include/tket/Circuit/Multiplexor.hpp [hunk 3]
```diff
@@ -96,6 +96,22 @@ class MultiplexorBox : public Box {
   ctrl_op_map_t op_map_;
 };
 
+struct GateSpec {
+  // TODO: this struct is a little confused as only CX needs qubit, only
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply tket/include/tket/Circuit/Multiplexor.hpp:100]: Yeah that makes so much more sense - will do it

[the remark this answers is not in the document]
Comment (sjdilkes) [reply tket/include/tket/Circuit/Multiplexor.hpp:100]: added

**Review by yao-cqc — APPROVED** [review]
Looks good! Just need to bump the tket lib version

**Review by yao-cqc — APPROVED** [review]

**Pull request merged**